File: apps/go/simulator/atomic.go
package simulator

import (
	"math"
	"math/rand"
	"sync/atomic"
)

// ElementInfo stores periodic table data for an element.
type ElementInfo struct {
	Symbol            string
	Name              string
	Group             int
	Period            int
	Electronegativity float64
}

// Elements is a partial periodic table.
var Elements = map[int]ElementInfo{
	1:  {"H", "Hydrogen", 1, 1, 2.20},
	2:  {"He", "Helium", 18, 1, 0.0},
	3:  {"Li", "Lithium", 1, 2, 0.98},
	4:  {"Be", "Beryllium", 2, 2, 1.57},
	5:  {"B", "Boron", 13, 2, 2.04},
	6:  {"C", "Carbon", 14, 2, 2.55},
	7:  {"N", "Nitrogen", 15, 2, 3.04},
	8:  {"O", "Oxygen", 16, 2, 3.44},
	9:  {"F", "Fluorine", 17, 2, 3.98},
	10: {"Ne", "Neon", 18, 2, 0.0},
	11: {"Na", "Sodium", 1, 3, 0.93},
	12: {"Mg", "Magnesium", 2, 3, 1.31},
	13: {"Al", "Aluminum", 13, 3, 1.61},
	14: {"Si", "Silicon", 14, 3, 1.90},
	15: {"P", "Phosphorus", 15, 3, 2.19},
	16: {"S", "Sulfur", 16, 3, 2.58},
	17: {"Cl", "Chlorine", 17, 3, 3.16},
	18: {"Ar", "Argon", 18, 3, 0.0},
	19: {"K", "Potassium", 1, 4, 0.82},
	20: {"Ca", "Calcium", 2, 4, 1.00},
	26: {"Fe", "Iron", 8, 4, 1.83},
}

// ElectronShell represents an electron shell with quantum numbers.
type ElectronShell struct {
	N            int // Principal quantum number
	MaxElectrons int
	Electrons    int
}

// Full returns true if the shell is full.
func (s *ElectronShell) Full() bool {
	return s.Electrons >= s.MaxElectrons
}

// atomIDGen is the global atom ID counter.
var atomIDGen atomic.Int64

// Atom represents an atom with protons, neutrons, and electron shells.
type Atom struct {
	AtomicNumber     int
	MassNumber       int
	ElectronCount    int
	Position         [3]float64
	Velocity         [3]float64
	Shells           []ElectronShell
	Bonds            []int64 // IDs of bonded atoms
	ID               int64
	IonizationEnergy float64
}

// NewAtom creates a new atom with the given atomic number and optional mass number.
func NewAtom(atomicNumber, massNumber int) *Atom {
	if massNumber == 0 {
		massNumber = atomicNumber * 2
		if atomicNumber == 1 {
			massNumber = 1
		}
	}
	a := &Atom{
		AtomicNumber:  atomicNumber,
		MassNumber:    massNumber,
		ElectronCount: atomicNumber, // neutral atom
		ID:            atomIDGen.Add(1),
		Bonds:         make([]int64, 0),
	}
	a.buildShells()
	a.computeIonizationEnergy()
	return a
}

func (a *Atom) buildShells() {
	a.Shells = nil
	remaining := a.ElectronCount
	for i, maxE := range ElectronShells {
		if remaining <= 0 {
			break
		}
		e := remaining
		if e > maxE {
			e = maxE
		}
		a.Shells = append(a.Shells, ElectronShell{
			N:            i + 1,
			MaxElectrons: maxE,
			Electrons:    e,
		})
		remaining -= e
	}
}

func (a *Atom) computeIonizationEnergy() {
	if len(a.Shells) == 0 || a.Shells[len(a.Shells)-1].Electrons == 0 {
		a.IonizationEnergy = 0.0
		return
	}
	n := float64(a.Shells[len(a.Shells)-1].N)
	innerE := 0
	for i := 0; i < len(a.Shells)-1; i++ {
		innerE += a.Shells[i].Electrons
	}
	zEff := float64(a.AtomicNumber - innerE)
	a.IonizationEnergy = 13.6 * zEff * zEff / (n * n)
}

// Symbol returns the element symbol.
func (a *Atom) Symbol() string {
	if elem, ok := Elements[a.AtomicNumber]; ok {
		return elem.Symbol
	}
	return "?"
}

// Electronegativity returns the electronegativity of the atom.
func (a *Atom) Electronegativity() float64 {
	if elem, ok := Elements[a.AtomicNumber]; ok {
		return elem.Electronegativity
	}
	return 1.0
}

// Charge returns the net charge of the atom.
func (a *Atom) Charge() int {
	return a.AtomicNumber - a.ElectronCount
}

// ValenceElectrons returns valence electrons.
func (a *Atom) ValenceElectrons() int {
	if len(a.Shells) == 0 {
		return 0
	}
	return a.Shells[len(a.Shells)-1].Electrons
}

// IsNobleGas returns true if the outermost shell is full.
func (a *Atom) IsNobleGas() bool {
	if len(a.Shells) == 0 {
		return false
	}
	return a.Shells[len(a.Shells)-1].Full()
}

// CanBondWith checks if bonding is possible with another atom.
func (a *Atom) CanBondWith(other *Atom) bool {
	if a.IsNobleGas() || other.IsNobleGas() {
		return false
	}
	if len(a.Bonds) >= 4 || len(other.Bonds) >= 4 {
		return false
	}
	return true
}

// BondType determines bond type based on electronegativity difference.
func (a *Atom) BondType(other *Atom) string {
	diff := math.Abs(a.Electronegativity() - other.Electronegativity())
	if diff > 1.7 {
		return "ionic"
	} else if diff > 0.4 {
		return "polar_covalent"
	}
	return "covalent"
}

// BondEnergy calculates bond energy.
func (a *Atom) BondEnergy(other *Atom) float64 {
	btype := a.BondType(other)
	switch btype {
	case "ionic":
		return BondEnergyIonic
	case "polar_covalent":
		return (BondEnergyCovalent + BondEnergyIonic) / 2
	default:
		return BondEnergyCovalent
	}
}

// DistanceTo returns Euclidean distance to another atom.
func (a *Atom) DistanceTo(other *Atom) float64 {
	dx := a.Position[0] - other.Position[0]
	dy := a.Position[1] - other.Position[1]
	dz := a.Position[2] - other.Position[2]
	return math.Sqrt(dx*dx + dy*dy + dz*dz)
}

// HasBondWith returns true if a bond exists with the given atom ID.
func (a *Atom) HasBondWith(id int64) bool {
	for _, b := range a.Bonds {
		if b == id {
			return true
		}
	}
	return false
}

// RemoveBond removes a bond by atom ID.
func (a *Atom) RemoveBond(id int64) {
	for i, b := range a.Bonds {
		if b == id {
			a.Bonds = append(a.Bonds[:i], a.Bonds[i+1:]...)
			return
		}
	}
}

// AtomicSystem is a collection of atoms with interactions.
type AtomicSystem struct {
	Atoms        []*Atom
	Temperature  float64
	BondsFormed  int
	BondsBroken  int
	rng          *rand.Rand
}

// NewAtomicSystem creates a new atomic system.
func NewAtomicSystem(rng *rand.Rand) *AtomicSystem {
	return &AtomicSystem{
		Atoms:       make([]*Atom, 0, 128),
		Temperature: TRecombination,
		rng:         rng,
	}
}

// Recombination captures free electrons into ions when T < T_recombination.
func (as *AtomicSystem) Recombination(field *QuantumField) []*Atom {
	if as.Temperature > TRecombination {
		return nil
	}

	var newAtoms []*Atom
	var protons []*Particle
	var electrons []*Particle
	for _, p := range field.Particles {
		switch p.Type {
		case ParticleProton:
			protons = append(protons, p)
		case ParticleElectron:
			electrons = append(electrons, p)
		}
	}

	for _, proton := range protons {
		if len(electrons) == 0 {
			break
		}
		electron := electrons[len(electrons)-1]
		electrons = electrons[:len(electrons)-1]

		atom := NewAtom(1, 1)
		atom.Position = proton.Position
		atom.Velocity = proton.Momentum

		newAtoms = append(newAtoms, atom)
		as.Atoms = append(as.Atoms, atom)

		field.removeParticle(proton)
		field.removeParticle(electron)
	}

	return newAtoms
}

// Nucleosynthesis forms heavier elements through nuclear fusion.
func (as *AtomicSystem) Nucleosynthesis(protons, neutrons int) []*Atom {
	var newAtoms []*Atom

	// Helium-4: 2 protons + 2 neutrons
	for protons >= 2 && neutrons >= 2 {
		he := NewAtom(2, 4)
		he.Position = [3]float64{as.rng.NormFloat64() * 10, as.rng.NormFloat64() * 10, as.rng.NormFloat64() * 10}
		newAtoms = append(newAtoms, he)
		as.Atoms = append(as.Atoms, he)
		protons -= 2
		neutrons -= 2
	}

	// Remaining protons become hydrogen
	for i := 0; i < protons; i++ {
		h := NewAtom(1, 1)
		h.Position = [3]float64{as.rng.NormFloat64() * 10, as.rng.NormFloat64() * 10, as.rng.NormFloat64() * 10}
		newAtoms = append(newAtoms, h)
		as.Atoms = append(as.Atoms, h)
	}

	return newAtoms
}

// StellarNucleosynthesis forms heavier elements in stellar cores.
func (as *AtomicSystem) StellarNucleosynthesis(temperature float64) []*Atom {
	var newAtoms []*Atom
	if temperature < 1e3 {
		return newAtoms
	}

	// Collect heliums
	var heliumIdx []int
	for i, a := range as.Atoms {
		if a.AtomicNumber == 2 {
			heliumIdx = append(heliumIdx, i)
		}
	}

	// Triple-alpha process: 3 He -> C
	for len(heliumIdx) >= 3 && as.rng.Float64() < 0.01 {
		// Remove 3 heliums (iterate from end to preserve indices)
		toRemove := heliumIdx[len(heliumIdx)-3:]
		heliumIdx = heliumIdx[:len(heliumIdx)-3]
		// Remove in reverse order to keep indices valid
		for i := len(toRemove) - 1; i >= 0; i-- {
			idx := toRemove[i]
			as.Atoms = append(as.Atoms[:idx], as.Atoms[idx+1:]...)
		}

		carbon := NewAtom(6, 12)
		carbon.Position = [3]float64{as.rng.NormFloat64() * 5, as.rng.NormFloat64() * 5, as.rng.NormFloat64() * 5}
		newAtoms = append(newAtoms, carbon)
		as.Atoms = append(as.Atoms, carbon)
	}

	// C + He -> O
	for as.rng.Float64() < 0.02 {
		ci, hi := -1, -1
		for i, a := range as.Atoms {
			if a.AtomicNumber == 6 && ci == -1 {
				ci = i
			}
			if a.AtomicNumber == 2 && hi == -1 {
				hi = i
			}
		}
		if ci == -1 || hi == -1 {
			break
		}

		pos := as.Atoms[ci].Position
		// Remove in correct order
		if ci > hi {
			as.Atoms = append(as.Atoms[:ci], as.Atoms[ci+1:]...)
			as.Atoms = append(as.Atoms[:hi], as.Atoms[hi+1:]...)
		} else {
			as.Atoms = append(as.Atoms[:hi], as.Atoms[hi+1:]...)
			as.Atoms = append(as.Atoms[:ci], as.Atoms[ci+1:]...)
		}

		oxygen := NewAtom(8, 16)
		oxygen.Position = pos
		newAtoms = append(newAtoms, oxygen)
		as.Atoms = append(as.Atoms, oxygen)
	}

	// O + He -> N (simplified fusion chain)
	if as.rng.Float64() < 0.005 {
		oi, hi := -1, -1
		for i, a := range as.Atoms {
			if a.AtomicNumber == 8 && oi == -1 {
				oi = i
			}
			if a.AtomicNumber == 2 && hi == -1 {
				hi = i
			}
		}
		if oi != -1 && hi != -1 {
			pos := as.Atoms[oi].Position
			if oi > hi {
				as.Atoms = append(as.Atoms[:oi], as.Atoms[oi+1:]...)
				as.Atoms = append(as.Atoms[:hi], as.Atoms[hi+1:]...)
			} else {
				as.Atoms = append(as.Atoms[:hi], as.Atoms[hi+1:]...)
				as.Atoms = append(as.Atoms[:oi], as.Atoms[oi+1:]...)
			}

			nitrogen := NewAtom(7, 14)
			nitrogen.Position = pos
			newAtoms = append(newAtoms, nitrogen)
			as.Atoms = append(as.Atoms, nitrogen)
		}
	}

	return newAtoms
}

// ElementCounts returns atom counts by element symbol.
func (as *AtomicSystem) ElementCounts() map[string]int {
	counts := make(map[string]int)
	for _, a := range as.Atoms {
		counts[a.Symbol()]++
	}
	return counts
}

File: apps/php/simulator/Atomic.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/Constants.php';
require_once __DIR__ . '/Quantum.php';

/**
 * Atomic physics simulation.
 *
 * Models atoms with electron shells, ionization, chemical bonding potential,
 * and the periodic table. Atoms emerge from the quantum/nuclear level
 * when temperature drops below recombination threshold.
 */

/**
 * An electron shell (energy level) within an atom.
 *
 * Models the capacity and occupancy of a principal quantum shell,
 * supporting addition and removal of electrons.
 */
class ElectronShell
{
    /** @var int Current number of electrons in this shell. */
    public int $electrons;

    /**
     * Create an electron shell.
     *
     * @param int $n            Principal quantum number (1, 2, 3, ...).
     * @param int $maxElectrons Maximum electrons this shell can hold.
     * @param int $electrons    Initial electron count.
     */
    public function __construct(
        public readonly int $n,
        public readonly int $maxElectrons,
        int $electrons = 0,
    ) {
        $this->electrons = $electrons;
    }

    /**
     * Check if this shell is fully occupied.
     *
     * @return bool True if electron count equals or exceeds max capacity.
     */
    public function isFull(): bool
    {
        return $this->electrons >= $this->maxElectrons;
    }

    /**
     * Check if this shell has no electrons.
     *
     * @return bool True if electron count is zero.
     */
    public function isEmpty(): bool
    {
        return $this->electrons === 0;
    }

    /**
     * Add one electron to this shell if not full.
     *
     * @return bool True if the electron was added, false if shell is full.
     */
    public function addElectron(): bool
    {
        if (!$this->isFull()) {
            $this->electrons++;
            return true;
        }
        return false;
    }

    /**
     * Remove one electron from this shell if not empty.
     *
     * @return bool True if an electron was removed, false if shell is empty.
     */
    public function removeElectron(): bool
    {
        if (!$this->isEmpty()) {
            $this->electrons--;
            return true;
        }
        return false;
    }
}

/**
 * An atom with protons, neutrons, and electron shells.
 *
 * Models atomic structure including electron configuration, ionization,
 * electronegativity, chemical bonding potential, and inter-atom distance.
 * Atoms are the building blocks created during recombination and nucleosynthesis.
 */
class Atom
{
    /** @var int Auto-incrementing counter for unique atom IDs. */
    private static int $idCounter = 0;

    /** @var int Unique identifier for this atom instance. */
    public readonly int $atomId;
    public int $massNumber;
    public int $electronCount;

    /** @var float[] */
    public array $position;

    /** @var float[] */
    public array $velocity;

    /** @var ElectronShell[] */
    public array $shells = [];

    /** @var int[] atom IDs of bonded atoms */
    public array $bonds = [];

    public float $ionizationEnergy = 0.0;

    /**
     * Create a new atom.
     *
     * @param int        $atomicNumber Number of protons (determines element).
     * @param int        $massNumber   Protons + neutrons (0 = auto-calculated).
     * @param int        $electronCount Number of electrons (0 = neutral atom).
     * @param float[]|null $position   3D position vector [x, y, z].
     * @param float[]|null $velocity   3D velocity vector [vx, vy, vz].
     */
    public function __construct(
        public readonly int $atomicNumber,
        int $massNumber = 0,
        int $electronCount = 0,
        ?array $position = null,
        ?array $velocity = null,
    ) {
        self::$idCounter++;
        $this->atomId = self::$idCounter;
        $this->position = $position ?? [0.0, 0.0, 0.0];
        $this->velocity = $velocity ?? [0.0, 0.0, 0.0];

        $this->massNumber = $massNumber ?: ($atomicNumber === 1 ? 1 : $atomicNumber * 2);
        $this->electronCount = $electronCount ?: $atomicNumber; // Neutral atom

        $this->buildShells();
        $this->computeIonizationEnergy();
    }

    /**
     * Reset the atom ID counter (useful for testing).
     */
    public static function resetIdCounter(): void
    {
        self::$idCounter = 0;
    }

    /**
     * Build electron shell configuration based on current electron count.
     */
    private function buildShells(): void
    {
        $this->shells = [];
        $remaining = $this->electronCount;
        foreach (ELECTRON_SHELLS as $i => $maxE) {
            if ($remaining <= 0) {
                break;
            }
            $electrons = min($remaining, $maxE);
            $this->shells[] = new ElectronShell(
                n: $i + 1,
                maxElectrons: $maxE,
                electrons: $electrons,
            );
            $remaining -= $electrons;
        }
    }

    private function computeIonizationEnergy(): void
    {
        if (empty($this->shells) || end($this->shells)->isEmpty()) {
            $this->ionizationEnergy = 0.0;
            return;
        }
        $lastShell = end($this->shells);
        $n = $lastShell->n;
        $innerElectrons = 0;
        foreach ($this->shells as $s) {
            if ($s !== $lastShell) {
                $innerElectrons += $s->electrons;
            }
        }
        $zEff = $this->atomicNumber - $innerElectrons;
        // Hydrogen-like approximation: E = 13.6 * Z_eff^2 / n^2
        $this->ionizationEnergy = 13.6 * $zEff ** 2 / $n ** 2;
    }

    /**
     * Get the chemical symbol for this element (e.g., 'H', 'He', 'C').
     *
     * @return string The element symbol, or 'E{Z}' for unknown elements.
     */
    public function symbol(): string
    {
        $elem = ELEMENTS[$this->atomicNumber] ?? null;
        return $elem ? $elem[0] : "E{$this->atomicNumber}";
    }

    /**
     * Get the full element name (e.g., 'Hydrogen', 'Carbon').
     *
     * @return string The element name, or 'Element-{Z}' for unknown elements.
     */
    public function name(): string
    {
        $elem = ELEMENTS[$this->atomicNumber] ?? null;
        return $elem ? $elem[1] : "Element-{$this->atomicNumber}";
    }

    /**
     * Get the Pauling electronegativity of this element.
     *
     * @return float Electronegativity value (0.0 for noble gases, ~4.0 for fluorine).
     */
    public function electronegativity(): float
    {
        $elem = ELEMENTS[$this->atomicNumber] ?? null;
        return $elem ? $elem[4] : 1.0;
    }

    /**
     * Get the net ionic charge (protons minus electrons).
     *
     * @return int Positive for cations, negative for anions, 0 for neutral.
     */
    public function charge(): int
    {
        return $this->atomicNumber - $this->electronCount;
    }

    /**
     * Get the number of electrons in the outermost shell.
     *
     * @return int Valence electron count.
     */
    public function valenceElectrons(): int
    {
        if (empty($this->shells)) {
            return 0;
        }
        return end($this->shells)->electrons;
    }

    /**
     * Get the number of additional electrons needed to fill the outermost shell.
     *
     * @return int Number of electrons needed (0 if shell is full).
     */
    public function needsElectrons(): int
    {
        if (empty($this->shells)) {
            return 0;
        }
        $shell = end($this->shells);
        return $shell->maxElectrons - $shell->electrons;
    }

    /**
     * Check if this atom has a full outer electron shell (noble gas configuration).
     *
     * @return bool True if the outermost shell is fully occupied.
     */
    public function isNobleGas(): bool
    {
        if (empty($this->shells)) {
            return false;
        }
        return end($this->shells)->isFull();
    }

    /**
     * Check if this atom is an ion (non-zero charge).
     *
     * @return bool True if the atom has gained or lost electrons.
     */
    public function isIon(): bool
    {
        return $this->charge() !== 0;
    }

    /**
     * Remove one electron from the atom (ionization).
     *
     * @return bool True if successful, false if no electrons remain.
     */
    public function ionize(): bool
    {
        if ($this->electronCount > 0) {
            $this->electronCount--;
            $this->buildShells();
            $this->computeIonizationEnergy();
            return true;
        }
        return false;
    }

    /**
     * Add one electron to the atom (electron capture).
     *
     * @return bool Always returns true.
     */
    public function captureElectron(): bool
    {
        $this->electronCount++;
        $this->buildShells();
        $this->computeIonizationEnergy();
        return true;
    }

    /**
     * Check if this atom can form a chemical bond with another atom.
     *
     * Noble gases and atoms with 4 existing bonds cannot form new bonds.
     *
     * @param Atom $other The other atom to potentially bond with.
     * @return bool True if bonding is possible.
     */
    public function canBondWith(Atom $other): bool
    {
        if ($this->isNobleGas() || $other->isNobleGas()) {
            return false;
        }
        if (count($this->bonds) >= 4 || count($other->bonds) >= 4) {
            return false;
        }
        return true;
    }

    /**
     * Determine the bond type based on electronegativity difference.
     *
     * @param Atom $other The other atom in the bond.
     * @return string 'ionic', 'polar_covalent', or 'covalent'.
     */
    public function bondType(Atom $other): string
    {
        $diff = abs($this->electronegativity() - $other->electronegativity());
        if ($diff > 1.7) {
            return 'ionic';
        } elseif ($diff > 0.4) {
            return 'polar_covalent';
        }
        return 'covalent';
    }

    /**
     * Calculate the bond energy for a bond with another atom.
     *
     * @param Atom $other The other atom in the bond.
     * @return float Bond energy in eV equivalent.
     */
    public function bondEnergy(Atom $other): float
    {
        return match ($this->bondType($other)) {
            'ionic'          => BOND_ENERGY_IONIC,
            'polar_covalent' => (BOND_ENERGY_COVALENT + BOND_ENERGY_IONIC) / 2,
            default          => BOND_ENERGY_COVALENT,
        };
    }

    /**
     * Calculate the Euclidean distance to another atom.
     *
     * @param Atom $other The other atom.
     * @return float Distance in simulation units.
     */
    public function distanceTo(Atom $other): float
    {
        $sum = 0.0;
        for ($i = 0; $i < 3; $i++) {
            $sum += ($this->position[$i] - $other->position[$i]) ** 2;
        }
        return sqrt($sum);
    }
}

class AtomicSystem
{
    /** @var Atom[] */
    public array $atoms = [];

    public float $temperature;

    /** @var Particle[] */
    public array $freeElectrons = [];

    public int $bondsFormed = 0;
    public int $bondsBroken = 0;

    public function __construct(float $temperature = T_RECOMBINATION)
    {
        $this->temperature = $temperature;
    }

    /** Capture free electrons into ions when T < T_recombination. */
    public function recombination(QuantumField $field): array
    {
        if ($this->temperature > T_RECOMBINATION) {
            return [];
        }

        $newAtoms = [];
        $protons = [];
        $electrons = [];

        foreach ($field->particles as $p) {
            if ($p->particleType === ParticleType::Proton) {
                $protons[] = $p;
            } elseif ($p->particleType === ParticleType::Electron) {
                $electrons[] = $p;
            }
        }

        foreach ($protons as $proton) {
            if (empty($electrons)) {
                break;
            }
            $electron = array_pop($electrons);
            $atom = new Atom(
                atomicNumber: 1,
                massNumber: 1,
                position: $proton->position,
                velocity: $proton->momentum,
            );
            $newAtoms[] = $atom;
            $this->atoms[] = $atom;

            // Remove from quantum field
            $idx = array_search($proton, $field->particles, true);
            if ($idx !== false) {
                array_splice($field->particles, $idx, 1);
            }
            $idx = array_search($electron, $field->particles, true);
            if ($idx !== false) {
                array_splice($field->particles, $idx, 1);
            }
        }

        return $newAtoms;
    }

    /** Form heavier elements through nuclear fusion. */
    public function nucleosynthesis(int $protons = 0, int $neutrons = 0): array
    {
        $newAtoms = [];

        // Helium-4: 2 protons + 2 neutrons
        while ($protons >= 2 && $neutrons >= 2) {
            $he = new Atom(
                atomicNumber: 2,
                massNumber: 4,
                position: [self::gauss(0, 10), self::gauss(0, 10), self::gauss(0, 10)],
            );
            $newAtoms[] = $he;
            $this->atoms[] = $he;
            $protons -= 2;
            $neutrons -= 2;
        }

        // Remaining protons become hydrogen
        for ($i = 0; $i < $protons; $i++) {
            $h = new Atom(
                atomicNumber: 1,
                massNumber: 1,
                position: [self::gauss(0, 10), self::gauss(0, 10), self::gauss(0, 10)],
            );
            $newAtoms[] = $h;
            $this->atoms[] = $h;
        }

        return $newAtoms;
    }

    /** Form heavier elements in stellar cores. */
    public function stellarNucleosynthesis(float $temperature): array
    {
        $newAtoms = [];

        if ($temperature < 1e3) {
            return $newAtoms;
        }

        $heliums = [];
        foreach ($this->atoms as $a) {
            if ($a->atomicNumber === 2) {
                $heliums[] = $a;
            }
        }

        // Triple-alpha process: 3 He -> C
        while (count($heliums) >= 3 && self::rand() < 0.01) {
            for ($i = 0; $i < 3; $i++) {
                $he = array_pop($heliums);
                $idx = array_search($he, $this->atoms, true);
                if ($idx !== false) {
                    array_splice($this->atoms, $idx, 1);
                }
            }

            $carbon = new Atom(
                atomicNumber: 6,
                massNumber: 12,
                position: [self::gauss(0, 5), self::gauss(0, 5), self::gauss(0, 5)],
            );
            $newAtoms[] = $carbon;
            $this->atoms[] = $carbon;
        }

        // C + He -> O
        $carbons = [];
        $heliums = [];
        foreach ($this->atoms as $a) {
            if ($a->atomicNumber === 6) {
                $carbons[] = $a;
            } elseif ($a->atomicNumber === 2) {
                $heliums[] = $a;
            }
        }
        while (!empty($carbons) && !empty($heliums) && self::rand() < 0.02) {
            $c = array_pop($carbons);
            $he = array_pop($heliums);

            $idx = array_search($c, $this->atoms, true);
            if ($idx !== false) {
                array_splice($this->atoms, $idx, 1);
            }
            $idx = array_search($he, $this->atoms, true);
            if ($idx !== false) {
                array_splice($this->atoms, $idx, 1);
            }

            $oxygen = new Atom(
                atomicNumber: 8,
                massNumber: 16,
                position: $c->position,
            );
            $newAtoms[] = $oxygen;
            $this->atoms[] = $oxygen;
        }

        // O + He -> N (simplified chain)
        $oxygens = [];
        $heliums = [];
        foreach ($this->atoms as $a) {
            if ($a->atomicNumber === 8) {
                $oxygens[] = $a;
            } elseif ($a->atomicNumber === 2) {
                $heliums[] = $a;
            }
        }
        if (!empty($oxygens) && !empty($heliums) && self::rand() < 0.005) {
            $o = $oxygens[0];
            $he = $heliums[0];

            $idx = array_search($o, $this->atoms, true);
            if ($idx !== false) {
                array_splice($this->atoms, $idx, 1);
            }
            $idx = array_search($he, $this->atoms, true);
            if ($idx !== false) {
                array_splice($this->atoms, $idx, 1);
            }

            $nitrogen = new Atom(
                atomicNumber: 7,
                massNumber: 14,
                position: $o->position,
            );
            $newAtoms[] = $nitrogen;
            $this->atoms[] = $nitrogen;
        }

        return $newAtoms;
    }

    /** Try to form a chemical bond between two atoms. */
    public function attemptBond(Atom $a1, Atom $a2): bool
    {
        if (!$a1->canBondWith($a2)) {
            return false;
        }

        $dist = $a1->distanceTo($a2);
        $bondDist = 2.0;

        if ($dist > $bondDist * 3) {
            return false;
        }

        $energyBarrier = $a1->bondEnergy($a2);
        $thermalEnergy = K_B * $this->temperature;
        if ($thermalEnergy > 0) {
            $prob = exp(-$energyBarrier / $thermalEnergy);
        } else {
            $prob = $dist < $bondDist ? 1.0 : 0.0;
        }

        if (self::rand() < $prob) {
            $a1->bonds[] = $a2->atomId;
            $a2->bonds[] = $a1->atomId;
            $this->bondsFormed++;
            return true;
        }
        return false;
    }

    /** Break a bond due to thermal energy. */
    public function breakBond(Atom $a1, Atom $a2): bool
    {
        if (!in_array($a2->atomId, $a1->bonds)) {
            return false;
        }

        $energyBarrier = $a1->bondEnergy($a2);
        $thermalEnergy = K_B * $this->temperature;

        if ($thermalEnergy > $energyBarrier * 0.5) {
            $prob = exp(-$energyBarrier / ($thermalEnergy + 1e-20));
            if (self::rand() < $prob) {
                $a1->bonds = array_values(array_diff($a1->bonds, [$a2->atomId]));
                $a2->bonds = array_values(array_diff($a2->bonds, [$a1->atomId]));
                $this->bondsBroken++;
                return true;
            }
        }
        return false;
    }

    /** Cool the atomic system. */
    public function cool(float $factor = 0.999): void
    {
        $this->temperature *= $factor;
    }

    /** Count atoms by element. */
    public function elementCounts(): array
    {
        $counts = [];
        foreach ($this->atoms as $a) {
            $sym = $a->symbol();
            $counts[$sym] = ($counts[$sym] ?? 0) + 1;
        }
        return $counts;
    }

    public function toSnapshot(): array
    {
        return [
            'temperature'   => $this->temperature,
            'atom_count'    => count($this->atoms),
            'elements'      => $this->elementCounts(),
            'bonds_formed'  => $this->bondsFormed,
            'bonds_broken'  => $this->bondsBroken,
        ];
    }

    private static function rand(): float
    {
        return mt_rand() / (mt_getrandmax() + 1);
    }

    private static function gauss(float $mean, float $stddev): float
    {
        $u1 = max(1e-10, mt_rand() / (mt_getrandmax() + 1));
        $u2 = mt_rand() / (mt_getrandmax() + 1);
        return $mean + $stddev * sqrt(-2 * log($u1)) * cos(2 * M_PI * $u2);
    }
}
